File: day1_php/src/FrequencyCalculator.php
<?php
namespace App;

class FrequencyCalculator
{
    private $frequencies;
    private $results;
    private $occurrences;
    private $firstFrequencyReachedTwice;

    public function __construct()
    {
        $this->frequencies = array();
        $this->results = array();
        $this->occurrences = array(1);
        $this->firstFrequencyReachedTwice = null;
    }

    /**
     * @return integer
     */
    public function getFrequency()
    {
        if (count($this->results) == 0 ) {
            return 0;
        } else {
            return $this->results[count($this->results) - 1];
        }
    }

    /**
     * @param mixed $frequency
     */
    public function setFrequency($frequency)
    {
        $this->results[] = $frequency;
    }

    /**
     * @param mixed $frequency
     */
    public function addFrequency($frequency)
    {
        $this->frequencies[] = $frequency;

        $result = null;
        if (count($this->results) == 0 ) {
            $result = 0 + $frequency;
        } else {
            $result = $this->results[count($this->results)-1] + $frequency;
        }
        $this->results[] = $result;

        $occurrence = 1;
        for ($i=count($this->results)-2; $i>=0; $i--) {
            if ($this->results[$i] == $result) {
                $occurrence = $this->occurrences[$i] + 1;
                if (null === $this->firstFrequencyReachedTwice) {
                    $this->firstFrequencyReachedTwice = $i;
                    print_r('first key: ' . $i);
                }
            }
        }
        $this->occurrences[] = $occurrence;
    }

    /**
     * @return array
     */
    public function getResults()
    {
        return $this->results;
    }

    /**
     * @return integer
     */
    public function getFirstFrequencyReachedTwice() {
        $ok = false;
        $result = null;

        $frequenciesToRepeat = count($this->frequencies);

        $nbIter = 0;
        while ($this->firstFrequencyReachedTwice === null) {
            for ($i=0; $i<$frequenciesToRepeat; $i++) {
                $this->addFrequency($this->frequencies[$i]);
            }
//            foreach ($this->occurrences as $key => $occurrence) {
//                if ($occurrence > 1) {
//                    $result = $this->results[$key];
//                    $ok = true;
//                    break; // break foreach
//                }
//            }
            $nbIter++;
            print_r(sprintf('iteration : %s ', $nbIter));
        }

        print_r($this->results);
        return $this->results[$this->firstFrequencyReachedTwice];
    }
}

File: day1_java/src/main/java/com/sebf/aoc/day1/FrequencyCalculator.java
package com.sebf.aoc.day1;

import java.util.ArrayList;

public class FrequencyCalculator {
    private ArrayList<Long> frequencies;
    private ArrayList<Long> results;
    private ArrayList<Long> occurrences;
    private Integer firstFrequencyReachedTwice;

    public FrequencyCalculator() {
        frequencies = new ArrayList<Long>();
        results = new ArrayList<Long>();
        occurrences = new ArrayList<Long>();
        occurrences.add(1L);
        firstFrequencyReachedTwice = null;
    }

    public Long getFrequency() {
        if (results.size() == 0) {
            return 0L;
        } else {
            return results.get(results.size() - 1);
        }
    }

    public void setFrequency(long frequency) {
        results.add(frequency);
    }

    public void addFrequency(long frequency) {
        frequencies.add(frequency);

        Long result;
        if (results.size() == 0) {
            result = frequency;
        } else {
            result = results.get(results.size() - 1) + frequency;
        }
        results.add(result);

        long occurrence = 1;
        for (int i = results.size() - 2; i >= 0; i--) {
            if (result.equals(results.get(i))) {
                occurrence = occurrences.get(i) + 1;
                if (null == firstFrequencyReachedTwice) {
                    firstFrequencyReachedTwice = i;
                }
            }
        }
        occurrences.add(occurrence);
    }

    public ArrayList<Long> getResults() {
        return results;
    }

    public Long getFirstFrequencyReachedTwice() {
        boolean ok = false;
        Long result = null;

        int frequenciesToRepeat = frequencies.size();

        long nbIter = 0;
        while (firstFrequencyReachedTwice == null) {
            for (int i = 0; i < frequenciesToRepeat; i++) {
                addFrequency(frequencies.get(i));
            }
            nbIter++;
            System.out.println(String.format("iteration : %s ", nbIter));
        }

        return results.get(firstFrequencyReachedTwice);
    }
}
